internal/proxy/proxy_test.go

```diff
@@ -0,0 +1,170 @@
+		"Password is escaped":                          {http: "http://username:p@$$:w0rd@example.com:8080"},
+		"Escaped password is not escaped again":        {http: "http://username:p%40$$%3Aw0rd@example.com:8080"},
+		"Domain username is escaped":                   {http: `http://EXAMPLE\bobsmith:p@$$:w0rd@example.com:8080`},
+		"Domain username without password is escaped":  {http: `http://EXAMPLE\bobsmith@example.com:8080`},
+		"Escaped domain username is not escaped again": {http: `http://EXAMPLE%5Cbobsmith@example.com:8080`},
+
+		"Proxy files not created when ran with dry-run": {dryRun: true},
+
+		// Error cases
+		"Error when environment config file is read-only":    {existingPerms: map[string]os.FileMode{envConfigPath: 0444}, prevContents: map[string]string{envConfigPath: "something"}, wantErr: true},
+		"Error when environment directory is a file":         {existingDirs: []string{"etc/"}, prevContents: map[string]string{filepath.Dir(envConfigPath): "something"}, wantErr: true},
+		"Error when environment directory is not executable": {existingPerms: map[string]os.FileMode{filepath.Dir(envConfigPath): 0444}, wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.existingDirs == nil {
+				tc.existingDirs = []string{filepath.Dir(envConfigPath)}
+			}
+
+			root := t.TempDir()
+			for _, p := range tc.existingDirs {
+				err := os.MkdirAll(filepath.Join(root, p), 0700)
+				require.NoError(t, err, "Setup: Couldn't create %s", p)
+			}
+
+			for p, c := range tc.prevContents {
+				err := os.WriteFile(filepath.Join(root, p), []byte(c), 0600)
+				require.NoError(t, err, "Setup: Couldn't write previous contents to %q", p)
+
+				err = os.Chtimes(filepath.Join(root, p), time.Now().UTC(), initialTime)
+				require.NoError(t, err, "Setup: Couldn't change mtime for %q", p)
+			}
+
+			for file, perms := range tc.existingPerms {
+				testutils.Chmod(t, filepath.Join(root, file), perms)
+			}
+
+			ctx := context.Background()
+			if tc.dryRun {
+				ctx = context.WithValue(ctx, proxy.DryRun, true)
+			}
+
+			p, err := proxy.New(ctx, tc.http, tc.https, tc.ftp, tc.socks, tc.noProxy, tc.mode, proxy.WithRoot(root))
+			require.NoError(t, err, "Setup: New failed but shouldn't have")
+
+			err = p.Apply(ctx)
+			if tc.wantErr {
+				require.Error(t, err, "Apply should have failed but didn't")
+				return
+			}
+			require.NoError(t, err, "Apply failed but shouldn't have")
+
+			testutils.CompareTreesWithFiltering(t, root, testutils.GoldenPath(t), testutils.Update())
+			for _, file := range tc.wantUnchangedFiles {
+				fi, err := os.Stat(filepath.Join(root, file))
+				require.NoError(t, err, "Setup: Failed to stat proxy config file")
+				require.Equal(t, initialTime, fi.ModTime().UTC(), "Proxy config file was modified")
```

Comment:
Here, I would more write what we expect from Equal, like "Proxy config file should not be modified"